AX: comboboxes with aria-activedescendant but no aria-owns no longer work

minorninth opened the pull request:
#### 481028cb765bb767493e48b704e8a64f5907ae37
<pre>
AX: comboboxes with aria-activedescendant but no aria-owns no longer work
<a href="https://bugs.webkit.org/show_bug.cgi?id=263979">https://bugs.webkit.org/show_bug.cgi?id=263979</a>
<a href="https://rdar.apple.com/117747058">rdar://117747058</a>

Reviewed by Tyler Wilcock.

WebKit should fire a notification when aria-activedescendant changes on
a combobox, even when the combobox doesn&apos;t use aria-owns or aria-controls.

* LayoutTests/accessibility/aria-combobox-no-owns-expected.txt: Added.
* LayoutTests/accessibility/aria-combobox-no-owns.html: Added.
* LayoutTests/platform/glib/accessibility/aria-combobox-no-owns-expected.txt: Added.

Canonical link: <a href="https://commits.webkit.org/270182@main">https://commits.webkit.org/270182@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/48b62d68f9f209858dddd08f33cbf8ed408a52f5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24741 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3286 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25994 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26859 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22713 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25010 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/4960 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/723 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23064 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24986 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2336 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21359 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27442 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2043 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22290 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28453 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22566 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22632 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26263 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1977 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/287 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3273 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/22040 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5934 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2429 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2335 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/accessibility/aria-combobox-no-owns.html:
```diff
@@ -0,0 +1,50 @@
+<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN">
+<html>
+<head>
+<script src="../resources/js-test-pre.js"></script>
```
You can replace `<script src="../resources/js-test-pre.js"></script>` with `<script src="../resources/js-test.js"></script>`, and then remove the js-test-post.js include below (js-test.js does both).

twilco commented on LayoutTests/accessibility/aria-combobox-no-owns.html:
```diff
@@ -0,0 +1,50 @@
+<script src="../resources/js-test-pre.js"></script>
+<script src="../resources/accessibility-helper.js"></script>
+</head>
+<body id="body">
```
I think this id is unused (though I know we have many other tests that do this)

twilco commented on LayoutTests/accessibility/aria-combobox-no-owns.html:
```diff
@@ -0,0 +1,50 @@
+</div>
+
+<p id="description"></p>
+<div id="console"></div>
```
With js-test.js, these two elements aren't necessary.

twilco commented on LayoutTests/accessibility/aria-combobox-no-owns.html:
```diff
@@ -0,0 +1,50 @@
+<div id="console"></div>
+
+<script>
+    description("A combobox should still support aria-activedescendant even if it doesn't use aria-owns.");
```
Rather than using `description`, our new layout tests would use:

```
var output = "A combobox should still support aria-activedescendant even if it doesn't use aria-owns.\n\n";
```

And rather than using `debug` and `shouldBe` throughout the test, we would do something like:

```
// debug("list has " + list.selectedChildrenCount + " selected children");
output += `list has ${list.selectedChildrenCount} selected children\n`;

// shouldBe("list.selectedChildrenCount", "0");
// This expect function is defined in accessibility-helpers.js
output += expect("list.selectedChildrenCount", "0");
```

And then at the end of the test, we print the output:

```
debug(output);
finishJSTest();
```

The reason for this change of style is because `description`, `debug`, and `shouldBe` write to the DOM every time, which we've found can significantly slow test execution. With this new style, we only write to the DOM once.

A good example to follow is:

https://github.com/WebKit/WebKit/blob/eb3708646ee56c088e9dd24ac306e975391f949b/LayoutTests/accessibility/url-test.html

minorninth replied:
Thanks for pointing me to a good example to follow! I'll use that as a template for all new tests.

twilco commented on LayoutTests/accessibility/aria-combobox-no-owns.html:
```diff
@@ -0,0 +1,50 @@
+        window.jsTestIsAsync = true;
+        var selectedChildrenChangeCount = 0;
+
+        window.accessibilityController.addNotificationListener(function (target, notification) {
```
We might want to `accessibilityController.removeNotificationListener` right before `finishJSTest()` to clean this up. I'd hope this happens automatically between tests, but just to be safe might be good to include.

twilco commented on LayoutTests/accessibility/aria-combobox-no-owns.html:
```diff
@@ -0,0 +1,50 @@
+        document.getElementById("combobox").setAttribute("aria-activedescendant", "item1");
+        var listitem1 = accessibilityController.accessibleElementById("item1");
+        setTimeout(async function() {
+            debug("list has " + list.selectedChildrenCount + " selected children");
```
We can use `expectAsync` here to ensure we wait for the AX tree to be updated after the DOM change:

```
output += await expectAsync("list.selectedChildrenCount", "1");
```

twilco reviewed (commented):
Thank you for fixing this! Just style comments on the test.

Commit:
AX: comboboxes with aria-activedescendant but no aria-owns no longer work
https://bugs.webkit.org/show_bug.cgi?id=263979
rdar://117747058

Reviewed by Tyler Wilcock.

WebKit should fire a notification when aria-activedescendant changes on
a combobox, even when the combobox doesn't use aria-owns or aria-controls.

* LayoutTests/accessibility/aria-combobox-no-owns-expected.txt: Added.
* LayoutTests/accessibility/aria-combobox-no-owns.html: Added.
* LayoutTests/platform/glib/accessibility/aria-combobox-no-owns-expected.txt: Added.

Canonical link: https://commits.webkit.org/270182@main
